pycbc/inference/models/hierarchical.py

```diff
@@ -599,3 +604,316 @@ def _loglikelihood(self):
+            if lbl not in kwargs['primary_lbl']:
+                for param in wfparam_map[lbl]:
+                    subcp.set('static_params', param.subname, 'REPLACE')
+
+            # save the vitual config file to disk for later check
+            with open('%s.ini' % lbl, 'w', encoding='utf-8') as file:
+                subcp.write(file)
+
+            # initialize
+            submodel = read_from_config(subcp)
+
+            if lbl not in kwargs['primary_lbl']:
+                # similar to the standard hierarchical model,
+                # move the static params back to variable if sub-model
+                # doesn't need marginalization
+                for p in vparam_map[lbl]:
+                    submodel.static_params.pop(p.subname)
+                submodel.variable_params = tuple(p.subname
+                                                 for p in vparam_map[lbl])
+                # similar to the standard hierarchical model,
+                # remove the waveform transform parameters if sub-model
+                # doesn't need marginalization
+                for p in wfparam_map[lbl]:
+                    submodel.static_params.pop(p.subname)
+            submodels[lbl] = submodel
+            logging.info("")
+
+        # remove all marginalized parameters from the top-level model's
+        # `variable_params` and `prior` sections
+        # here we ignore `coa_phase`, because if it's been marginalized,
+        # it will not be listed in `variable_params` and `prior` sections
+        primary_model = submodels[kwargs['primary_lbl'][0]]
+        marginalized_params = primary_model.marginalize_vector_params.copy()
+        marginalized_params.pop('logw_partial')
+        marginalized_params = list(marginalized_params.keys())
+        # this may also include 'f_ref', 'f_lower', 'approximant',
+        # but doesn't matter
+        marginalized_params += list(primary_model.static_params.keys())
+        for p in primary_model.static_params.keys():
+            p_full = '%s__%s' % (kwargs['primary_lbl'][0], p)
+            if p_full not in cp['static_params']:
+                cp['static_params'][p_full] = "%s" % \
+                    primary_model.static_params[p]
+
+        for section in cp.sections():
+            if 'prior-' in section:
+                p = section.split('-')[-1]
+                if p in marginalized_params:
+                    cp['variable_params'].pop(p)
+                    cp.pop(section)
+
+        # now load the model
+        logging.info("Loading joint_primary_marginalized model")
+        return super(HierarchicalModel, cls).from_config(
+                cp, submodels=submodels, **kwargs)
+
+    def reconstruct(self, seed=None):
+        """ Reconstruct marginalized parameters by using the primary
+        model's reconstruct method, total_loglr and others_lognl.
+        """
+        rec = self.primary_model.reconstruct(
+                seed=seed, set_loglr=self.total_loglr())
```

Comment:
How is this supposed to work? You are just calling self.total_loglr and then passing in the return? Don't you need to pass in a *function* that can be called in the reconstruct method?

Reply:
Now I almost exactly follow the structure of the original `recntruct`, define a `get_loglr` function to call my `total_loglr()`, my `self.total_loglr()` corresponding to `self.loglr`.

Reply:
I forgot the difference between `set_loglr=get_loglr` and `set_loglr=get_loglr()`, now I know we need to use  `set_loglr=get_loglr` if we want to pass a _function_ to it.